components/scream/src/physics/p3/p3_functions_droplet_self_coll_impl.hpp

```diff
@@ -0,0 +1,35 @@
+#ifndef P3_FUNCTIONS_DROPLET_SELF_COLL_IMPL_HPP
+#define P3_FUNCTIONS_DROPLET_SELF_COLL_IMPL_HPP
+
+#include "p3_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 droplet self collection function. Clients should NOT
+ * #include this file, but include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::droplet_self_collection(const Spack& rho, const Spack& inv_rho,
+                          const Spack& qc_incld, const Spack& mu_c,
+                          const Spack& nu, const Spack& ncautc, Spack& ncslf)
+{
+  constexpr Scalar qsmall = C::QSMALL;
+
+  const auto qc_not_small = (qc_incld >= qsmall);
+  if (qc_not_small.any()) {
+    // Khroutdinov and Kogan (2000)
+    ncslf.set(qc_not_small, 0.0);
+
+    ncslf.set(ncslf == 0, 0);
```

Comment:
I don't think you need this line.

Reply:
Should we just replace with this, then?

```
ncslf.set(0);
```

I was just trying to be conservative and do what the Fortran code does. Those two statements technically leave alone entries that are smaller than qsmall but nonzero.

Reply:
We want to mimic the fortran behavior, so keep the qc_not_small set.

The second line seems redundant. If ncslf is zero, set ncslf to zero?

Reply:
Yeah, I guess I assumed something else was going on. We shouldn't be using == with floating point types in the first place, if we really want to be careful. I'll ax it.